src/priv_appendable_data.rs

```diff
@@ -0,0 +1,487 @@
+        // Count valid previous_owner_signatures and refuse if quantity is not enough
+
+        let check_all_keys = |&sig| {
+            owner_keys.iter()
+                .any(|pub_key| sign::verify_detached(&sig, &data, pub_key))
+        };
+
+        if self.previous_owner_signatures
+            .iter()
+            .filter(|&sig| check_all_keys(sig))
+            .count() < (owner_keys.len() / 2 + owner_keys.len() % 2) {
+            return Err(RoutingError::NotEnoughSignatures);
+        }
+        Ok(())
+    }
+
+    fn data_to_sign(&self) -> Result<Vec<u8>, RoutingError> {
+        // Seems overkill to use serialisation here, but done to ensure cross platform signature
+        // handling is OK
+        let sd = SerialisablePrivAppendableData {
+            name: self.name,
+            previous_owner_keys: &self.previous_owner_keys,
+            current_owner_keys: &self.current_owner_keys,
+            version: self.version.to_string().as_bytes().to_vec(),
+            filter: &self.filter,
+            encrypt_key: &self.encrypt_key,
+            deleted_data: &self.deleted_data,
+        };
+
+        serialise(&sd).map_err(From::from)
+    }
+
+    /// Adds a signature with the given `secret_key` to the `previous_owner_signatures` and returns
+    /// the number of signatures that are still required. If more than 50% of the previous owners
+    /// have signed, 0 is returned and validation is complete.
+    pub fn add_signature(&mut self, secret_key: &SecretKey) -> Result<usize, RoutingError> {
+        let data = try!(self.data_to_sign());
+        let sig = sign::sign_detached(&data, secret_key);
+        self.previous_owner_signatures.push(sig);
+        let owner_keys = if self.previous_owner_keys.is_empty() {
+            &self.current_owner_keys
+        } else {
+            &self.previous_owner_keys
+        };
+        Ok(((owner_keys.len() / 2) + 1).saturating_sub(self.previous_owner_signatures.len()))
+    }
+
+    /// Overwrite any existing signatures with the new signatures provided.
+    pub fn replace_signatures(&mut self, new_signatures: Vec<Signature>) {
+        self.previous_owner_signatures = new_signatures;
+    }
+
+    /// Get the data
+    pub fn get_data(&self) -> &BTreeSet<PrivAppendedData> {
+        &self.data
+    }
+
+    /// Get the previous owner keys
+    pub fn get_previous_owner_keys(&self) -> &Vec<PublicKey> {
+        &self.previous_owner_keys
+    }
+
+    /// Get the version
+    pub fn get_version(&self) -> u64 {
+        self.version
+    }
+
+    /// Get the current owner keys
+    pub fn get_owner_keys(&self) -> &Vec<PublicKey> {
+        &self.current_owner_keys
+    }
+
+    /// Get previous owner signatures
+    pub fn get_previous_owner_signatures(&self) -> &Vec<Signature> {
+        &self.previous_owner_signatures
+    }
+
+    /// Return data size.
+    pub fn payload_size(&self) -> usize {
+        self.data.len() * SERIALISED_APPENDED_DATA_SIZE
```

Comment:
But that should probably be multiplied by the size of `PrivAppendedData` not `AppendedData`, I think.